README
Project taille

A simple tool that may prints aout the sizes of some structures such
as intergers for instance.


ChangeLog
05.08.2008 Olivier Delhomme <olivierdelhomme@gmail.com>
	   * Project skelleton


autoclean.sh
#!/bin/bash
#
# Author : Olivier Delhomme
# Licence : GNU GPL v2
#
if [ -f Makefile ];
 then
    make distclean
fi;
rm -f aclocal.m4
echo "rm -f aclocal.m4"
rm -fr autom4te.cache
echo "rm -fr autom4te.cache"
rm -f config.guess
echo "rm -f config.guess"
rm -f config.sub
echo "rm -f config.sub"
rm -f depcomp
echo "rm -f depcomp"
rm -f INSTALL
echo "rm -f INSTALL"
rm -f install-sh
echo "rm -f install-sh"
rm -f ltmain.sh
echo "rm -f ltmain.sh"
rm -f missing
echo "rm -f missing"
rm -f src/*~ src/*#
echo "rm -f src/*~ src/*#"
rm -f *~ *#
echo "rm -f *~ *#"
rm -f Makefile.in src/Makefile.in
echo "rm -f Makefile.in src/Makefile.in"
rm -f configure
echo "rm -f configure"
rm -f config.status config.log
echo "rm -f config.status config.log"
rm -f libtool
echo "rm -f libtool"
find -name *~ -exec rm {} \;
echo "find -name *~ -exec rm {} \;"
find -name *.gcd[ao] -exec rm -f {} \;
echo "find -name *.gcd[ao] -exec rm -f {} \;"
find -name *.gcov -exec rm -f {} \;
echo "find -name *.gcov -exec rm -f {} \;"
find -name *.gcno -exec rm -f {} \;
echo "find -name *.gcno -exec rm -f {} \;"
rm -f intltool-extract.in
echo "rm -f intltool-extract.in"
rm -f intltool-merge.in
echo "rm -f intltool-merge.in"
rm -f intltool-update.in
echo "rm -f intltool-update.in"
rm -f intltool-extract
echo "rm -f intltool-extract"
rm -f intltool-merge
echo "rm -f intltool-merge"
rm -f intltool-update
echo "rm -f intltool-update"
rm -f src/gmon.out
echo "rm -f src/gmon.out"


src/taille.c
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: t; c-basic-offset: 4 -*- */
/*
 *  taille.c
 *  taille - a simple tool that prints out sizes.
 *
 *  (C) Copyright 2008 - 2015 Olivier Delhomme
 *  e-mail : olivier.delhomme@free.fr
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2, or  (at your option) 
 *  any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY;  without even the implied warranty of
 *  MERCHANTABILITY  or  FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. 
 */

#include "taille.h"

static gboolean version(void);
static gboolean usage(int status);
static gboolean manage_command_line_options(Options *opt, int argc, char ** argv);
static void init_international_languages(void);
static void print_int_types(void);


/**
 *  prints program name, version, author, date and licence
 *  to the standard output
 */
static gboolean version(void)
{
	fprintf (stdout, _("%s, %s - %s - Version %s - License %s\n"), ProgName, ProgAuthor, ProgDate, ProgVersion, _("GPL V2 or higher"));
	return TRUE;
}


/**
 *  Function that informs the user about the command line options
 *  available with taille
 */
static gboolean usage(int status)
{
	if (status == 0)
		{
			fprintf (stderr, 
					 _("Try `%s --help' for more information.\n"), ProgName);
			return FALSE;
		}
	else
		{
			version();
			fprintf(stdout, _("\n%s is a tool that prints struct sizes"), ProgName);
			fprintf(stdout, _("\nUsage:\n  %s [options]\n"), ProgName);
			fprintf(stdout, _("\nOptions:\n\
  -h, --help\tThis help.\n\
  -v, --version\tProgram version information.\n"));
			return TRUE;
		}
}


/**
 *  Manages all the command line options and populates the
 *  Options *opt structure accordingly
 */
static gboolean manage_command_line_options(Options *opt, int argc, char ** argv)
{
	int exit_value = TRUE;
	int c = 0;

	while ((c = getopt_long (argc, argv, "vh", long_options, NULL)) != -1)
		{
			switch (c)
				{
				case 0:				/* long options handler */
					break;
	  
				case 'v':
					exit_value = version();
					opt->usage = TRUE; 	/* We do not want to continue after */
					break;
	 
				case 'h':
					exit_value = usage(1);
					opt->usage = TRUE;
					break;

				default:
					exit_value = usage(0);
					opt->usage = TRUE;
				}
		}

	if (optind < argc) /* filename */
		{
			exit_value = usage(0);
			opt->usage = TRUE;
		}

	return exit_value;
}


/**
 *  Inits internationalization domain for taille project
 */
static void init_international_languages(void)
{
    gchar *result = NULL;
    gchar *codeset = NULL;
    gchar *text_domain = NULL;

    setlocale(LC_ALL, "");
    result = bindtextdomain(GETTEXT_PACKAGE, LOCALE_DIR);
    codeset = bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
    text_domain = textdomain(GETTEXT_PACKAGE);

	if (DEBUG == TRUE)
		{
			fprintf(stdout, "Gettext package : %s\n", GETTEXT_PACKAGE);

			if (result != NULL)
				{
					fprintf(stdout, "Bindtextdomain : %s\n", result);
				}

			if (codeset != NULL)
				{
					fprintf(stdout, "Code set : %s\n", codeset);
				}

			if (text_domain != NULL)
				{
					fprintf(stdout, "Text domain : %s\n", text_domain);
				}
		}
}




/**
 *  Prints int types
 */
static void print_int_types(void)
{
	unsigned long taille = 0;

	fprintf(stdout, _("Integer types:\n"));
	
	taille = sizeof(short int);
	fprintf(stdout, "\tshort int    : %ld (%ld bits)\n", taille, taille*8);

	taille = sizeof(int);	
	fprintf(stdout, "\tint          : %ld (%ld bits)\n", taille, taille*8);
	
	taille = sizeof(long int);
	fprintf(stdout, "\tlong int     : %ld (%ld bits)\n", taille, taille*8);

    taille = sizeof(long long int);
	fprintf(stdout, "\tlong long int: %ld (%ld bits)\n", taille, taille*8); 
}



/**
 *  main program
 *  options :
 *   --version
 *   --help
 */
int main (int argc, char ** argv) 
{  
	Options *opt;   /* A structure to manage the command line options  */
	gboolean exit_value = TRUE;

	opt = (Options *) g_malloc0(sizeof(Options));
	opt->usage = FALSE;
	
	init_international_languages();

	/* Command line options evaluation */
	exit_value = manage_command_line_options(opt, argc, argv);
	
	if (opt->usage != TRUE)
		{
			print_int_types();
		}

	g_free(opt);

	return exit_value;
}


src/include/taille.h
/* -*- indent-tabs-mode: t; tab-width: 4; c-basic-offset: 4; -*- */
/*
   taille.h
   taille project.

   (C) Copyright 2008 - 2015 Olivier Delhomme
   e-mail : olivier.delhomme@free.fr
 
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.           
*/

#ifndef _TAILLE_H_
#define _TAILLE_H_

/* Configuration from ./configure script */
#include "config.h"

#include <stdio.h>
#include <getopt.h>
#include <glib.h>
#include <glib/gi18n.h>

#define ProgName "taille"
#define ProgVersion "0.0.1"
#define ProgDate "05.08.2008"
#define ProgAuthor "Olivier Delhomme <olivier.delhomme@free.fr>"

typedef struct
{
	char usage;        /* To know if we displayed the usage message */
} Options;

static struct option const long_options[] =
{
	{"version", no_argument, NULL, 'v'},
	{"help", no_argument, NULL, 'h'},
	{NULL, 0, NULL, 0} 
};


#endif /* _TAILLE_H_ */
